lib/Dialect/Secret/Conversions/SecretToBGV/SecretToBGV.cpp

```diff
@@ -41,59 +45,81 @@ namespace mlir::heir {
 // TODO(#536): Integrate a general library to compute appropriate prime moduli
 // given any number of bits.
-FailureOr<::mlir::heir::polynomial::RingAttr> getRlweRing(
-    MLIRContext *ctx, int coefficientModBits, int polyModDegree) {
+FailureOr<polynomial::RingAttr> getRlweRNSRing(MLIRContext *ctx,
+                                               int currentLevel,
+                                               int coefficientModBits,
+                                               int polyModDegree) {
   std::vector<::mlir::heir::polynomial::IntMonomial> monomials;
   monomials.emplace_back(1, polyModDegree);
   monomials.emplace_back(1, 0);
   auto result =
       ::mlir::heir::polynomial::IntPolynomial::fromMonomials(monomials);
   if (failed(result)) return failure();
   ::mlir::heir::polynomial::IntPolynomial xnPlusOne = result.value();
-  switch (coefficientModBits) {
-    case 29: {
-      auto type = IntegerType::get(ctx, 32);
-      APInt defaultMod(32, 463187969);
-      return ::mlir::heir::polynomial::RingAttr::get(
-          mod_arith::ModArithType::get(ctx, IntegerAttr::get(type, defaultMod)),
-          polynomial::IntPolynomialAttr::get(ctx, xnPlusOne));
-    }
-    default:
-      return failure();
+  // TODO: all 40 bit primes...
+  std::vector<int64_t> primes = {1095233372161, 1032955396097, 1005037682689,
+                                 998595133441,  972824936449,  959939837953};
+  SmallVector<Type, 4> modTypes;
+  for (int i = 0; i <= currentLevel; i++) {
+    auto type = IntegerType::get(ctx, 64);
+    modTypes.push_back(
+        mod_arith::ModArithType::get(ctx, IntegerAttr::get(type, primes[i])));
   }
+  auto rnsType = rns::RNSType::get(ctx, modTypes);
+  return ::mlir::heir::polynomial::RingAttr::get(
+      rnsType, polynomial::IntPolynomialAttr::get(ctx, xnPlusOne));
+}
+
+polynomial::RingAttr getRlweRNSRingWithLevel(polynomial::RingAttr ringAttr,
+                                             int level) {
+  auto rnsType = cast<rns::RNSType>(ringAttr.getCoefficientType());
+
+  auto newRnsType = rns::RNSType::get(
+      rnsType.getContext(), rnsType.getBasisTypes().take_front(level + 1));
+  return ::mlir::heir::polynomial::RingAttr::get(
+      newRnsType, ringAttr.getPolynomialModulus());
 }
 
 }  // namespace
 
-// Remove this class if no type conversions are necessary
 class SecretToBGVTypeConverter : public TypeConverter {
  public:
   SecretToBGVTypeConverter(MLIRContext *ctx,
                            ::mlir::heir::polynomial::RingAttr rlweRing) {
+    ring = rlweRing;
+
+    // actually isLegal/isSignatureLegal will always be true
+    // relying on addIllegalDialect<secret::SecretDialect>
+    // for checking function signature correctness
     addConversion([](Type type) { return type; });
 
     // Convert secret types to BGV ciphertext types
-    addConversion([ctx, this](secret::SecretType type) -> Type {
+    addTypeAttributeConversion([ctx, this](secret::SecretType type,
```

Comment:
I think this is the only "hacky" part of this PR: that you're implementing this function for something it's not intended to be used for, just because it happens to be a member of TypeConverter already and TypeAttr happens to be a legal return value. It's clever but I think we can do even better.

What if instead we had a little HEIR-specific class hierarchy:

```cpp
// in ConversionUtils.h
class ContextAwareTypeConverter : public TypeConverter {
  // Returns a TypeRange with the converted types of the values in the input range,
  // taking into account the context of the values (e.g., defining ops or uses).
  TypeRange convertValueRangeTypes(ValueRange values);

  // Returns a TypeRange with the converted types of the results of an op,
  // taking into account the context of the op when selecting the new type.
  TypeRange convertOpResultTypes(Operation *op, TypeRange resultTypes);
}

// in SecretToBGV.cpp
class SecretToBGVTypeConverter : ContextAwareTypeConverter {
  Type convertTypeWithAttr(Type type, Attribute *attr) override {
    // your single-type conversion in addTypeAttributeConversion

    // if attr is nullptr, fall back to the base class type converter
  }

  TypeRange convertValueRangeTypes(ValueRange values) override {
    // attribute extraction logic and invocation of convertTypeWithAttr
  }

  TypeRange convertOpResultTypes(Operation *op, TypeRange resultTypes) {
    // similar to convertValueRangeTypes, but the mgmt attr extracted from `op`.
  }
}
```

Then in the patterns in ConversionUtils.h, you can `dyn_cast` the input `TypeConverter` to `ContextAwareTypeConverter`, and if the cast succeeds you'd call its new function instead (otherwise rely on `OpAdaptor` as usual). This has the benefit that `SecretToBGVTypeConverter` could even contain a DataflowSolver instance as a member variable, and reference the dataflow analysis results directly at type conversion time (would that eliminate the need to persist attributes on the ops in the first place?).

Moreover, since `ConvertFuncWithMgmt` isn't used as part of the actual dialect conversion framework (just `walkAndApplyPatterns`), it can inherit from RewritePattern instead of `OpConversionPattern`, directly take a `ContextAwareTypeConverter` in its constructor, and avoid dyn-casts entirely, which would be a little simpler and less confusing to the reader.

Reply:
> HEIR-specific class hierarchy

I was thinking that SecretToBGVTypeconverter should not appear in ConversionUtils (as CKKS also uses it) so chose to misuse its interface. A middle abstract Typeconverter could help.

> This has the benefit that SecretToBGVTypeConverter could even contain a DataflowSolver instance as a member variable, and reference the dataflow analysis results directly at type conversion time (would that eliminate the need to persist attributes on the ops in the first place?).

I think we still face the problem of DataflowSolver getting invalidated during the conversion and has to re-run it for every possible middle state of the conversion, which could be hard.

The current analysis only works after `mlir-to-secret-arithmetic` but before `secret-distribute-generic` so that things get simple, if we are analysing it at funcOp level things are going to be complex (transfer state for both GenericOp{arith::MulIOP} and bgv::MulOp}), where inside analysis we will check whether secret::GenericOp contains only one op; this is similar to the `persist attribute` technique but put more complexity on to the analysis instead of the conversion pass.

Reply:
Also the Secretnessanalysis mainly depends on the SecretType to work (so that should be used before `secret-to-scheme`), but the SecretType will be converted away during this conversion and the analysis has to deal with such complexity by analysing LWE types to give

Reply:
I still maintain the MgmtAttr approach as it is cleaner for debugging (debuging into analysis is quite hard). I think an Analysis based TypeConverter can be done in the future, but that will touch too many analysis to makes things work (SecretAnalysis and all this PR's analyses, handling all the intermediate state of dialect lowering) and worth another PR. It has been already too many thing in this PR.